File: Practice/Project1/cpp/hospital.cpp
#include "hospital.h"
#include "patient.h"
#include <iostream>

using namespace std;

Hospital::Hospital()
{
    max_patients = 5;
}

void Hospital::menu()
{
    cout << "Enter your choice:" << endl
         << "1) Add new patient" << endl
         << "2) Print all patients" << endl
         << "3) Get next patient" << endl
         << "4) Exit" << endl;
}

void Hospital::addPatient()
{
    int specialization;
    string name;
    int status;

    if (patients.size() < max_patients)
    {
        cout << "Enter specialization (1-20): " << endl;
        cin >> specialization;
        cout << "Enter patient name: " << endl;
        cin >> name;
        cout << "Enter status (0 for regular, 1 for urgent): " << endl;
        cin >> status;
        patients.push_back(Patient(name, specialization, status));
        cout << "Patient added successfully." << endl;
    }

    else
    {
        cout << "Queue is full." << endl;
    }
}

void Hospital::printAllPatients()
{
    if (patients.empty())
    {
        cout << "No patients in the queue." << endl;
        return;
    }

    cout << "Total patients: " << patients.size() << endl;
    for (size_t i = 0; i < patients.size(); ++i)
    {
        Patient &patient = patients[i];
        cout << "Patient " << (i + 1) << ":" << endl;
        cout << "  Name: " << patient.getName() << endl;
        cout << "  Specialization: " << patient.getSpecialization() << endl;
        cout << "  Status: " << (patient.getStatus() == 1 ? "urgent" : "regular") << endl;
    }
}

void Hospital::getNextPatient()
{
    if (patients.empty())
    {
        cout << "No patients in the queue." << endl;
        return;
    }
    Patient nextPatient = patients[0];
    patients.erase(patients.begin());
    cout << "Next patient:" << endl;
    cout << "  Name: " << nextPatient.getName() << endl;
    cout << "  Specialization: " << nextPatient.getSpecialization() << endl;
    cout << "  Status: " << (nextPatient.getStatus() == 1 ? "urgent" : "regular") << endl;
}

void Hospital::loop()
{
    while (true)
    {
        menu();
        char choice;
        cin >> choice;
        switch (choice)
        {
        case '1':
            addPatient();
            break;
        case '2':
            printAllPatients();
            break;
        case '3':
            getNextPatient();
            break;
        case '4':
            return;
        default:
            cout << "Invalid choice. Please try again." << endl;
        }
    }
}

File: Practice/Project1/py/hospital.py
from py.patient import Patient


class Hospital:
    """
    This class represents a hospital system with a single queue for patients.

    Attributes:
        patients (list): List of Patient objects waiting in the queue.
        max_patients (int): Maximum number of patients allowed in the queue.
    """

    def __init__(self):
        """
        Initializes the hospital system with an empty queue and a maximum patient limit.
        """
        self.patients = []
        self.max_patients = 5

    def menu(self):
        """
        Displays a menu with options for adding, printing, getting the next patient, and exiting.
        """
        print("Enter your choice:")
        print("1) Add new patient")
        print("2) Print all patients")
        print("3) Get next patient")
        print("4) Exit")

    def add_patient(self):
        """
        Adds a new patient to the queue if there's space, handling user input and potential errors.
        """
        if len(self.patients) < self.max_patients:
            try:
                specialization = int(input("Enter specialization (1-20): "))
                name = input("Enter patient name: ")
                status = int(input("Enter status (0 for regular, 1 for urgent): "))
                self.patients.append(Patient(specialization, name, status))
                print("Patient added successfully.")
            except ValueError:
                print("Invalid input. Please enter valid integers.")
        else:
            print("Queue is full.")

    def print_all_patients(self):
        """
        Prints information about all patients in the queue.
        """
        print(f"Total patients: {len(self.patients)}")
        for i, patient in enumerate(self.patients, start=1):
            print(f"Patient {i}:")
            print(f"  Name: {patient.name}")
            print(f"  Specialization: {patient.specialization}")
            print(f"  Status: {'urgent' if patient.status else 'regular'}")

    def get_next_patient(self):
        """
        Retrieves and displays information about the next patient in the queue.
        """
        if self.patients:
            next_patient = self.patients.pop(0)
            print("Next patient:")
            print(f"  Name: {next_patient.name}")
            print(f"  Specialization: {next_patient.specialization}")
            print(f"  Status: {'urgent' if next_patient.status else 'regular'}")
        else:
            print("No patients in the queue.")

    def loop(self):
        """
        Runs a continuous loop displaying the menu and handling user choices.
        """
        while True:
            self.menu()
            choice = input()
            if choice == "1":
                self.add_patient()
            elif choice == "2":
                self.print_all_patients()
            elif choice == "3":
                self.get_next_patient()
            elif choice == "4":
                break
            else:
                print("Invalid choice. Please try again.")
